File: leetcode/226-invert-binary-tree/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        """
        classic approach: recursion

        Time  O(n)
        Space O(h)
        24 ms, faster than 26.24%
        """
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def invertTree(self, root):
        if root == None:
            return None
        left = self.invertTree(root.left)
        right = self.invertTree(root.right)
        root.left = right
        root.right = left
        return root

class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode

        classic approach: recursion

        Time  O(n)
        Space O(h)
        24 ms, faster than 26.24%
        """
        if root == None:
            return None
        temp = root.left
        root.left = self.invertTree(root.right)
        root.right = self.invertTree(temp)
        return root


class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode

        classic approach: iterative bfs

        Time  O(n)
        Space O(n)
        20 ms, faster than 69.85%
        """
        if root == None:
            return None
        q = []
        q.append(root)
        while len(q) > 0:
            head = q.pop(0)
            temp = head.left
            head.left = head.right
            head.right = temp
            if head.left != None:
                q.append(head.left)
            if head.right != None:
                q.append(head.right)
        return root

File: leetcode/226-invert-binary-tree/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	class approach: recursion

	Time  O(n)
	Space O(h)
	20 ms, faster than 69.85%
*/
func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	temp := root.Left
	root.Left = invertTree(root.Right)
	root.Right = invertTree(temp)
	return root
}

/*
	class approach: iterative bfs

	Time  O(n)
	Space O(1)
	0 ms, faster than 100.00%
*/
func invertTree1(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	q := []*TreeNode{}
	q = append(q, root)
	for len(q) > 0 {
		head := q[len(q)-1]
		q = q[:len(q)-1]
		head.Left, head.Right = head.Right, head.Left
		if head.Left != nil {
			q = append(q, head.Left)
		}
		if head.Right != nil {
			q = append(q, head.Right)
		}
	}
	return root
}

func main() {

}
